### ISSUE TITLE
[Affiliate] Upload metadata doc to IPFS

### ISSUE THREAD

**Issue by alfetopito** [gnosis/cowswap#942]
# Summary

Having the metadata doc (see #941), [persist it on IPFS](https://docs.ipfs.io/concepts/persistence/#persistence-versus-permanence).
Feel free to propose a lib for interfacing with IPFS or implement the interface if it's simple enough.
The returned hash will be persisted in another task.

# Requirements

- Upload json doc to IPFS
- Propose IPFS interface of your choice

# Out of scope

- Generating json doc (#941)
- Storing referral info (#945)

# Additional context

- Wireframes: https://www.figma.com/file/P0AIwKkIUpEZtLukkCUM94/CowSwap_-Affiliate?node-id=48%3A224
- Design doc: https://docs.google.com/document/d/1PsdWn9WPFf2BNeoJNtClyra8jNvWk_bJ-g33TbpUf8s/edit#

**Comment by alfetopito** [comment]
# Update 2021-09-20

IPFS is back!

The idea is still the same, but this time I won't go too much into implementation details on what order or which fn to write etc, as we agreed.

Overall idea, that could be split into different tasks as you see fit:

## 1. Build metadata json doc

Mostly implemented, need to adapt to latest model changes but I don't expect much work

## 2. Find a lib for interfacing with IPFS

Something like https://github.com/ipfs/js-ipfs probably

## 3. Pin uploaded content

@anxolin has worked with https://www.pinata.cloud/ before.
I'm personally not clear how that would work, so I'll leave the requirement and ~hide~ delegate the research ;)

## 4. Decode hash from CID

From the received CID, decode it and extract the multihash digest. Something like:

```
const CID = require("cids");
const multihashes = require("multihashes");

for (const cid of [
  "QmS7jokoXWKRZHnR6sGupAjziHceuYd2gQ2CFGQETpBcBa",
  "bafybeibyd727tfn6pw5qdiekmchybfdhvd6j6b7rz63h5agthkdfiwdn2e",
]) {
  const { digest } = multihashes.decode(new CID(cid).multihash);
  console.log(`0x${Buffer.from(digest).toString("hex")}`);
}
```

The following is here for context and comes from a thread with Nick, Anxo and I.

> CIDv0s are just 34 bytes of encoded base58 data, where the first 2 bytes are always 0x1220 (for "Sha-256" and "32 byte digest length") followed by the 32-byte digest.
> (lib for doing that) https://github.com/near/as-base58

> A CIDv0 is:
One variable length integer to encode the multihash (https://github.com/multiformats/multicodec/blob/master/table.csv)
For v0, only Sha-256 is accepted (single byte 0x12)
One variable length integer for the digest length
SInce only Sha-256 is accepted, the length is always 32 bytes - which encodes as 0x20
The digest
SInce only Sha-256 is accepted, its always 32 bytes longs
This is way all base58 CIDv0s always start with Qm. This is because the most significant bytes are always 0x1220. (edited) 

> Also, there is a JS CID library for this: https://github.com/multiformats/js-cid

> But, if I'm being perfectly honest, I think cid.length === 46 && cid.startsWith("Qm") is probably enough.
> Assuming the CID in frontend will come directly from the response of IPFS call, I don't think we need to verify anything.
If we are reading it from somewhere else, then yes
>In that case I agree. Although, it could be the case that your IPFS node returns CIDv1s instead right?
I'm not sure what happens there - does it implicitly convert from one to the other? As in, if you add a file and get a CIDv1, can you access it by its CIDv0 (assuming the same multihash type?)
Although from the docs:
- cid-version [int]: CID version. Defaults to 0 unless an option that depends on CIDv1 is passed. (experimental). Required: no.
- hash [string]: Hash function to use. Implies CIDv1 if not sha2-256. (experimental). Default: sha2-256. Required: no.
edit: docs link: https://docs.ipfs.io/reference/http/api/#getting-started (edited) 
> I would think that we need to verify is using the “implicit” stuff. Cause if that changes we will notice.
Although, now u mention this, the backend will already validate and if they have alerts, then we will also know this happened.

> But by the time the backend verifies this, the order has already been placed with the "bad" hash/CID.
If we can enforce it when uploading to IPFS as Nick pointed out with cid-version=0 and hash=sha2-256 (even if redundant), I would assume we are good.

> $ tr -dc A-Za-z0-9 </dev/urandom | head -c 32 > foo
$ cat foo
w6rI5Z8LCVNjL9isO01N8mKWAL2fFJdN
$ curl -s -F "file=@foo" http://localhost:5001/api/v0/add
{"Name":"foo","Hash":"QmS7jokoXWKRZHnR6sGupAjziHceuYd2gQ2CFGQETpBcBa","Size":"40"}
$ curl -s http://localhost:8080/ipfs/QmS7jokoXWKRZHnR6sGupAjziHceuYd2gQ2CFGQETpBcBa
w6rI5Z8LCVNjL9isO01N8mKWAL2fFJdN
$ ipfs cid format -v 1 -b base32 QmS7jokoXWKRZHnR6sGupAjziHceuYd2gQ2CFGQETpBcBa
bafybeibyd727tfn6pw5qdiekmchybfdhvd6j6b7rz63h5agthkdfiwdn2e
$ curl -s http://localhost:8080/ipfs/bafybeibyd727tfn6pw5qdiekmchybfdhvd6j6b7rz63h5agthkdfiwdn2e
w6rI5Z8LCVNjL9isO01N8mKWAL2fFJdN

> It looks like IPFS accepts CIDv0 or v1 as long as the multihash is the same.

> If you want to use a different hash (say sha3-512) then you have to re-api/v0/add.

> Hmm, I get that we can freely transform between CID v0 <-> CID v1
What does that effectively mean for us? We'll still be storing the 32 bytes hash decoded from the CID, right?

> We'll still be storing the 32 bytes hash decoded from the CID, right?
Exactly. But we don't really have to worry about the CID version.

> I think we still do, given that each is a different format, thus must be decoded in a different way to extract the hash, no?

> Ah, well kind of. I mean in theory yes, but the JS CID library supports both v0 and v1 (as I suspect most CID implementations do). Also the multihashes are the same regarless of the CID version.
So:
const CID = require("cids");
const multihashes = require("multihashes");

> for (const cid of [
  "QmS7jokoXWKRZHnR6sGupAjziHceuYd2gQ2CFGQETpBcBa",
  "bafybeibyd727tfn6pw5qdiekmchybfdhvd6j6b7rz63h5agthkdfiwdn2e",
]) {
  const { digest } = multihashes.decode(new CID(cid).multihash);
  console.log(`0x${Buffer.from(digest).toString("hex")}`);
}
Outputs
0x381ff5f995be7dbb01a08a608f809467a8fc9f07f1cfb67e80d33a8654586dd1
0x381ff5f995be7dbb01a08a608f809467a8fc9f07f1cfb67e80d33a8654586dd1

## 5. Return the hash

Well, maybe this is actually implied from the previous step.
